run examples/mnist failed with backend/simplego

the output as follows:
```

panic: failed to train model 4000 steps: Loop.RunSteps(4000): failed TrainStep(LoopStep=0): failed to execute train/eval step: failed building computation graph for mean metric "Moving Average Accuracy": in method ArgMinMax(): not implemented
```

--- a/backends/default/xla.go
+++ b/backends/default/xla.go
@@ -1,4 +1,7 @@
-//go:build !noxla
+//go:build linux && amd64 && !noxla
+
+// For now XLA is only supported for linux/amd64.
+// TODO: change when more platforms are supported (linux/arm64, darwin/amd64, darwin/arm64, etc.)
 
 package _default
 

--- a/backends/shapeinference/shapeinference.go
+++ b/backends/shapeinference/shapeinference.go
@@ -738,3 +738,24 @@ func SliceOp(operand shapes.Shape, starts, limits, strides []int) shapes.Shape {
 
 	return outputShape
 }
+
+// ArgMinMaxOp calculates the output shape for an ArgMinMax operation.
+// It will be the shape of the operand minus the reduce axis.
+func ArgMinMaxOp(operand shapes.Shape, axis int, outputDType dtypes.DType) shapes.Shape {
+	if !outputDType.IsInt() {
+		exceptions.Panicf("ArgMinMax outputDType must be an integer type, got %s", outputDType)
+	}
+	if !operand.DType.IsFloat() && !operand.DType.IsInt() {
+		exceptions.Panicf("ArgMinMax operand DType must be a floating point or integer type, got %s", operand)
+	}
+	if operand.IsScalar() {
+		exceptions.Panicf("ArgMinMax requires a non-scalar operand, got %s", operand)
+	}
+	if axis < 0 || axis >= operand.Rank() {
+		exceptions.Panicf("ArgMinMax axis %d is out of range for operand %s", axis, operand)
+	}
+	newDims := slices.Clone(operand.Dimensions)
+	newDims = slices.Delete(newDims, axis, axis+1)
+	output := shapes.Make(outputDType, newDims...)
+	return output
+}

--- a/backends/shapeinference/shapeinference_test.go
+++ b/backends/shapeinference/shapeinference_test.go
@@ -349,3 +349,42 @@ func TestSliceOp(t *testing.T) {
 	require.Panics(t, func() { SliceOp(operand, validStarts, []int{8, 6}, validStrides) }, // limit[1]=6 > dimSize[1]=5
 		"%s Error Case 10 Failed: Limit > dimSize", opName)
 }
+
+func TestArgMinMaxOp(t *testing.T) {
+	// --- Valid Cases ---
+
+	// Case 1: 1D tensor
+	operand1 := MS(F32, 10)
+	expected1 := MS(I32)
+	output1 := ArgMinMaxOp(operand1, 0, I32)
+	require.True(t, expected1.Equal(output1), "Valid Case 1 Failed: Expected %s, got %s", expected1, output1)
+
+	// Case 2: 2D tensor, single axis
+	operand2 := MS(F32, 5, 6)
+	expected2 := MS(I8, 5)
+	output2 := ArgMinMaxOp(operand2, 1, expected2.DType)
+	require.True(t, expected2.Equal(output2), "Valid Case 2 Failed: Expected %s, got %s", expected2, output2)
+
+	// Case 3: 3D tensor, multiple axes
+	operand3 := MS(F32, 4, 5, 6)
+	expected3 := MS(U64, 5, 6)
+	output3 := ArgMinMaxOp(operand3, 0, expected3.DType)
+	require.True(t, expected3.Equal(output3), "Valid Case 3 Failed: Expected %s, got %s", expected3, output3)
+
+	// --- Error Cases ---
+
+	// Error 1: Invalid operand DType
+	require.Panics(t, func() {
+		ArgMinMaxOp(shapes.Make(dtypes.InvalidDType, 10), 0, I32)
+	}, "Error Case 1 Failed: Invalid operand DType")
+
+	// Error 2: Invalid axis (out of bounds)
+	require.Panics(t, func() {
+		ArgMinMaxOp(operand1, 1, I32) // operand1 is rank 1, axis 1 invalid
+	}, "Error Case 2 Failed: Invalid axis (out of bounds)")
+
+	// Error 3: Negative axis
+	require.Panics(t, func() {
+		ArgMinMaxOp(operand2, -1, I32)
+	}, "Error Case 3 Failed: Negative axis")
+}

--- a/backends/simplego/capabilities.go
+++ b/backends/simplego/capabilities.go
@@ -18,7 +18,8 @@ var Capabilities = backends.Capabilities{
 		backends.OpTypeParameter: true,
 		backends.OpTypeConstant:  true,
 
-		// Graph nodes.
+		// Other operations:
+		backends.OpTypeArgMinMax:       true,
 		backends.OpTypeBroadcast:       true,
 		backends.OpTypeBroadcastInDim:  true,
 		backends.OpTypeConcatenate:     true,

--- a/backends/simplego/exec_special_ops.go
+++ b/backends/simplego/exec_special_ops.go
@@ -34,6 +34,7 @@ func init() {
 	nodeExecutors[backends.OpTypeScatterMin] = execScatter
 	nodeExecutors[backends.OpTypeScatterSum] = execScatter
 	nodeExecutors[backends.OpTypeSlice] = execSlice
+	nodeExecutors[backends.OpTypeArgMinMax] = execArgMinMax
 
 	// For nodes with multiple outputs:
 	multiOutputsNodeExecutors[backends.OpTypeRngBitGenerator] = execRngBitGenerator
@@ -1408,4 +1409,175 @@ func execRngBitGenerator(backend *Backend, node *Node, inputs []*Buffer, inputsO
 	stateFlat[0] = binary.LittleEndian.Uint64(rngState[4 : 4+8])
 	stateFlat[1] = binary.LittleEndian.Uint64(rngState[4+8 : 4+16])
 	return []*Buffer{state, rngData}
+}
+
+// execArgMinMax ====================================================================================================
+
+const MaxArgMinMaxReductionSize = 0x8000_0000
+
+// execArgMinMax is the executor function registered for backends.OpTypeArgMinMax.
+func execArgMinMax(backend *Backend, node *Node, inputs []*Buffer, inputsOwned []bool) *Buffer {
+	operand := inputs[0]
+	reduceAxis := node.data.(*argMinMaxNode).axis
+	isMin := node.data.(*argMinMaxNode).isMin
+	output := backend.getBuffer(node.shape.DType, node.shape.Size())
+	output.shape = node.shape
+
+	// There are 3 sizes to iterate over: before and after the reduceAxis, and the size (dimension) of the reduce axis itself.
+	operandDims := operand.shape.Dimensions
+	operandRank := len(operandDims)
+	suffixSize := 1
+	for axis := reduceAxis + 1; axis < operandRank; axis++ {
+		suffixSize *= operandDims[axis]
+	}
+	prefixSize := 1
+	for axis := range reduceAxis {
+		prefixSize *= operand.shape.Dimensions[axis]
+	}
+	reduceSize := operandDims[reduceAxis]
+	if reduceSize >= MaxArgMinMaxReductionSize {
+		// If we need larger, change buildArgMinMax to use int64 instead of int32.
+		exceptions.Panicf("ArgMaxMin implementation only supports reduction on dimensions < %d, got operand shaped %s and reduce axis is %d",
+			MaxArgMinMaxReductionSize, operand.shape, reduceAxis)
+	}
+
+	// Instantiate the function to copy over results from ints:
+	buildCopyIntsFn := argMinMaxCopyIntsDTypeMap.Get(output.shape.DType).(func(output *Buffer) func(flatIdx int, values []int32))
+	copyIntsFn := buildCopyIntsFn(output)
+
+	// Dispatch to the generic implementation based on DType.
+	argMinMaxFn := argMinMaxDTypeMap.Get(operand.shape.DType).(func(backend *Backend, operand *Buffer, copyIntsFn func(flatIdx int, values []int32), prefixSize, reduceSize, suffixSize int, isMin bool))
+	argMinMaxFn(backend, operand, copyIntsFn, prefixSize, reduceSize, suffixSize, isMin)
+	return output
+}
+
+var (
+	argMinMaxDTypeMap         = NewDTypeMap("ArgMinMaxRun")
+	argMinMaxCopyIntsDTypeMap = NewDTypeMap("ArgMinMaxCopyInts")
+)
+
+// buildArgMinMaxCopyIntsFn creates a "copyInts" function to copy the given values starting at the flatIdx to
+// the output buffer.
+func buildArgMinMaxCopyIntsFn[T PODIntegerConstraints](output *Buffer) func(flatIdx int, values []int32) {
+	outputFlat := output.flat.([]T)
+	return func(flatIdx int, values []int32) {
+		for _, value := range values {
+			outputFlat[flatIdx] = T(value)
+			flatIdx++
+		}
+	}
+}
+
+func execArgMinMaxGeneric[T PODNumericConstraints](
+	backend *Backend, operand *Buffer, copyIntsFn func(flatIdx int, values []int32), prefixSize, reduceSize, suffixSize int, isMin bool) {
+	operandFlat := operand.flat.([]T)
+
+	// Temporary data to store argMax results, so we can traverse the operand sequentially.
+	currentBestBuffer := backend.getBuffer(operand.shape.DType, suffixSize)
+	currentBest := currentBestBuffer.flat.([]T)
+	currentArgBestBuffer := backend.getBuffer(dtypes.Int32, suffixSize)
+	currentArgBest := currentArgBestBuffer.flat.([]int32)
+
+	outputFlatIdx := 0
+	operandFlatIdx := 0
+	for range prefixSize {
+		// Initialize current best with first element of reduce axis:
+		for suffixIdx := range suffixSize {
+			currentBest[suffixIdx] = operandFlat[operandFlatIdx]
+			currentArgBest[suffixIdx] = 0
+			operandFlatIdx++
+		}
+
+		// Iterate over the rest of the elements of reduce axis:
+		if isMin {
+			// ArgMin
+			for reduceIdx := 1; reduceIdx < reduceSize; reduceIdx++ {
+				for suffixIdx := range suffixSize {
+					operandValue := operandFlat[operandFlatIdx]
+					operandFlatIdx++
+					if operandValue < currentBest[suffixIdx] {
+						currentBest[suffixIdx] = operandValue
+						currentArgBest[suffixIdx] = int32(reduceIdx)
+					}
+				}
+			}
+		} else {
+			// ArgMax
+			for reduceIdx := 1; reduceIdx < reduceSize; reduceIdx++ {
+				for suffixIdx := range suffixSize {
+					operandValue := operandFlat[operandFlatIdx]
+					operandFlatIdx++
+					if operandValue > currentBest[suffixIdx] {
+						currentBest[suffixIdx] = operandValue
+						currentArgBest[suffixIdx] = int32(reduceIdx)
+					}
+				}
+			}
+		}
+
+		// Copy over the result of the whole suffix.
+		copyIntsFn(outputFlatIdx, currentArgBest)
+		outputFlatIdx += suffixSize
+	}
+	backend.putBuffer(currentBestBuffer)
+	backend.putBuffer(currentArgBestBuffer)
+}
+
+func init() {
+	argMinMaxDTypeMap.Register(dtypes.BFloat16, execArgMinMaxGenericBFloat16)
+}
+
+func execArgMinMaxGenericBFloat16(
+	backend *Backend, operand *Buffer, copyIntsFn func(flatIdx int, values []int32), prefixSize, reduceSize, suffixSize int, isMin bool) {
+	operandFlat := operand.flat.([]bfloat16.BFloat16)
+
+	// Temporary data to store argMax results, so we can traverse the operand sequentially.
+	currentBestBuffer := backend.getBuffer(operand.shape.DType, suffixSize)
+	currentBest := currentBestBuffer.flat.([]bfloat16.BFloat16)
+	currentArgBestBuffer := backend.getBuffer(dtypes.Int32, suffixSize)
+	currentArgBest := currentArgBestBuffer.flat.([]int32)
+
+	outputFlatIdx := 0
+	operandFlatIdx := 0
+	for range prefixSize {
+		// Initialize current best with first element of reduce axis:
+		for suffixIdx := range suffixSize {
+			currentBest[suffixIdx] = operandFlat[operandFlatIdx]
+			currentArgBest[suffixIdx] = 0
+			operandFlatIdx++
+		}
+
+		// Iterate over the rest of the elements of reduce axis:
+		if isMin {
+			// ArgMin
+			for reduceIdx := 1; reduceIdx < reduceSize; reduceIdx++ {
+				for suffixIdx := range suffixSize {
+					operandValue := operandFlat[operandFlatIdx].Float32()
+					if operandValue < currentBest[suffixIdx].Float32() {
+						currentBest[suffixIdx] = operandFlat[operandFlatIdx]
+						currentArgBest[suffixIdx] = int32(reduceIdx)
+					}
+					operandFlatIdx++
+				}
+			}
+		} else {
+			// ArgMax
+			for reduceIdx := 1; reduceIdx < reduceSize; reduceIdx++ {
+				for suffixIdx := range suffixSize {
+					operandValue := operandFlat[operandFlatIdx].Float32()
+					if operandValue > currentBest[suffixIdx].Float32() {
+						currentBest[suffixIdx] = operandFlat[operandFlatIdx]
+						currentArgBest[suffixIdx] = int32(reduceIdx)
+					}
+					operandFlatIdx++
+				}
+			}
+		}
+
+		// Copy over the result of the whole suffix.
+		copyIntsFn(outputFlatIdx, currentArgBest)
+		outputFlatIdx += suffixSize
+	}
+	backend.putBuffer(currentBestBuffer)
+	backend.putBuffer(currentArgBestBuffer)
 }

--- a/backends/simplego/exec_special_ops_test.go
+++ b/backends/simplego/exec_special_ops_test.go
@@ -543,3 +543,38 @@ func TestExecSpecialOps_RngBitsGenerator(t *testing.T) {
 		})
 	}
 }
+
+func TestExecSpecialOps_ArgMinMaxOp(t *testing.T) {
+	// Test Case 1: Simple 1D argmin
+	y0 := graph.ExecOnce(backend, func(x *graph.Node) *graph.Node {
+		return graph.ArgMin(x, 0)
+	}, []float32{3, 1, 4, 1, 5})
+	fmt.Printf("\ty0=%s\n", y0.GoStr())
+	require.Equal(t, int32(1), y0.Value())
+
+	// Test Case 2: 2D argmax along axis 1 (columns)
+	y1 := graph.ExecOnce(backend, func(x *graph.Node) *graph.Node {
+		return graph.ArgMax(x, 1)
+	}, [][]int32{{1, 2, 3}, {4, 1, 2}, {7, 8, 5}})
+	fmt.Printf("\ty1=%s\n", y1.GoStr())
+	require.Equal(t, []int32{2, 0, 1}, y1.Value())
+
+	// Test Case 3: 2D argmin along axis 0 (rows) with BFloat16
+	y2 := graph.ExecOnce(backend, func(x *graph.Node) *graph.Node {
+		return graph.ArgMin(x, 0)
+	}, [][]bfloat16.BFloat16{
+		{bf16(1), bf16(2)},
+		{bf16(-1), bf16(3)},
+		{bf16(4), bf16(-2)}})
+	fmt.Printf("\ty2=%s\n", y2.GoStr())
+	require.Equal(t, []int32{1, 2}, y2.Value())
+
+	// Test Case 4: 3D argmax with repeated values
+	y3 := graph.ExecOnce(backend, func(x *graph.Node) *graph.Node {
+		return graph.ArgMax(x, 1)
+	}, [][][]float32{
+		{{1, 2}, {1, 0}, {1, -1}},
+		{{4, 3}, {4, 5}, {4, 2}}})
+	fmt.Printf("\ty3=%s\n", y3.GoStr())
+	require.Equal(t, [][]int32{{0, 0}, {0, 1}}, y3.Value())
+}

--- a/backends/simplego/gen_register_dtypes.go
+++ b/backends/simplego/gen_register_dtypes.go
@@ -234,6 +234,28 @@ func init() {
 	sliceDTypeMap.RegisterIfNotSet(dtypes.BFloat16, execSliceGeneric[bfloat16.BFloat16])
 	sliceDTypeMap.RegisterIfNotSet(dtypes.Bool, execSliceGeneric[bool])
 
+	// DTypeMap: argMinMaxDTypeMap
+	argMinMaxDTypeMap.RegisterIfNotSet(dtypes.Int8, execArgMinMaxGeneric[int8])
+	argMinMaxDTypeMap.RegisterIfNotSet(dtypes.Int16, execArgMinMaxGeneric[int16])
+	argMinMaxDTypeMap.RegisterIfNotSet(dtypes.Int32, execArgMinMaxGeneric[int32])
+	argMinMaxDTypeMap.RegisterIfNotSet(dtypes.Int64, execArgMinMaxGeneric[int64])
+	argMinMaxDTypeMap.RegisterIfNotSet(dtypes.Uint8, execArgMinMaxGeneric[uint8])
+	argMinMaxDTypeMap.RegisterIfNotSet(dtypes.Uint16, execArgMinMaxGeneric[uint16])
+	argMinMaxDTypeMap.RegisterIfNotSet(dtypes.Uint32, execArgMinMaxGeneric[uint32])
+	argMinMaxDTypeMap.RegisterIfNotSet(dtypes.Uint64, execArgMinMaxGeneric[uint64])
+	argMinMaxDTypeMap.RegisterIfNotSet(dtypes.Float32, execArgMinMaxGeneric[float32])
+	argMinMaxDTypeMap.RegisterIfNotSet(dtypes.Float64, execArgMinMaxGeneric[float64])
+
+	// DTypeMap: argMinMaxCopyIntsDTypeMap
+	argMinMaxCopyIntsDTypeMap.RegisterIfNotSet(dtypes.Int8, buildArgMinMaxCopyIntsFn[int8])
+	argMinMaxCopyIntsDTypeMap.RegisterIfNotSet(dtypes.Int16, buildArgMinMaxCopyIntsFn[int16])
+	argMinMaxCopyIntsDTypeMap.RegisterIfNotSet(dtypes.Int32, buildArgMinMaxCopyIntsFn[int32])
+	argMinMaxCopyIntsDTypeMap.RegisterIfNotSet(dtypes.Int64, buildArgMinMaxCopyIntsFn[int64])
+	argMinMaxCopyIntsDTypeMap.RegisterIfNotSet(dtypes.Uint8, buildArgMinMaxCopyIntsFn[uint8])
+	argMinMaxCopyIntsDTypeMap.RegisterIfNotSet(dtypes.Uint16, buildArgMinMaxCopyIntsFn[uint16])
+	argMinMaxCopyIntsDTypeMap.RegisterIfNotSet(dtypes.Uint32, buildArgMinMaxCopyIntsFn[uint32])
+	argMinMaxCopyIntsDTypeMap.RegisterIfNotSet(dtypes.Uint64, buildArgMinMaxCopyIntsFn[uint64])
+
 	// DTypePairMap: convertDTypePairMap
 	convertDTypePairMap.RegisterIfNotSet(dtypes.Int8, dtypes.Int8, execConvertDTypeGeneric[int8, int8])
 	convertDTypePairMap.RegisterIfNotSet(dtypes.Int8, dtypes.Int16, execConvertDTypeGeneric[int8, int16])

--- a/backends/simplego/ops.go
+++ b/backends/simplego/ops.go
@@ -342,7 +342,34 @@ func (b *Builder) RngBitGenerator(stateOp backends.Op, shape shapes.Shape) (newS
 	return
 }
 
-// Unary Operations:
+type argMinMaxNode struct {
+	axis  int
+	isMin bool
+}
+
+// ArgMinMax calculates the "argmin" or "argmax" across an axis of the given input array x.
+// outputDType defines the output of the argmin/argmax, it doesn't need to be the same as the input.
+// It's a form of reduction on the given axis, and that axis goes away. So the rank of the result is one less than
+// the rank of x.
+// Examples:
+//
+//	ArgMinMax(x={{2, 0, 7}, {-3, 4, 2}}, axis=1, isMin=true) -> {1, 0}  // (it chooses the 0 and the -3)
+//	ArgMinMax(x={{2, 0, 7}, {-3, 4, 2}}, axis=0, isMin=false) -> {0, 1, 0} // (it choose the 2, 4 and 7)
+func (b *Builder) ArgMinMax(operandOp backends.Op, axis int, outputDType dtypes.DType, isMin bool) backends.Op {
+	opType := backends.OpTypeArgMinMax
+	operand := b.checkOps(opType.String(), operandOp)[0]
+	outputShape := shapeinference.ArgMinMaxOp(operand.shape, axis, outputDType)
+	node := b.newNode(opType, outputShape, operand)
+	node.data = &argMinMaxNode{
+		axis,
+		isMin,
+	}
+	return node
+}
+
+//======================================================================================================================
+// Unary Operations ----------------------------------------------------------------------------------------------------
+//======================================================================================================================
 
 // Neg implements the backends.Builder interface.
 func (b *Builder) Neg(operand backends.Op) backends.Op {

--- a/docs/CHANGELOG.md
+++ b/docs/CHANGELOG.md
@@ -5,7 +5,7 @@
 * Package `simplego`:
   * Fixed `Gather` of scalar values.
   * Fixed `Where` checking of shape.
-  * Added `Erf` op.
+  * New ops: `Erf`, `ArgMinMax`. 
 * Package `types/tensors`:
   * Removed dependency to `gopjrt/pjrt` -- otherwise we'll always need to install the C/C++ library.
 

--- a/docs/CHANGELOG.md
+++ b/docs/CHANGELOG.md
@@ -5,7 +5,9 @@
 * Package `simplego`:
   * Fixed `Gather` of scalar values.
   * Fixed `Where` checking of shape.
-  * New ops: `Erf`, `ArgMinMax`. 
+  * New ops: `Erf`, `ArgMinMax`.
+* Package `backends/default`:
+  * Only include XLA by default on linux/amd64 platforms.
 * Package `types/tensors`:
   * Removed dependency to `gopjrt/pjrt` -- otherwise we'll always need to install the C/C++ library.
 

--- a/examples/mnist/demo/demo.go
+++ b/examples/mnist/demo/demo.go
@@ -28,7 +28,7 @@ import (
 	"github.com/janpfeifer/must"
 	"k8s.io/klog/v2"
 
-	_ "github.com/gomlx/gomlx/backends/xla"
+	_ "github.com/gomlx/gomlx/backends/default"
 )
 
 // Example usage:  go run demo.go -train -checkpoint=cnn_triplet_hard_01 -set="model=cnn;loss=triplet;triplet_loss_mining_strategy=hard;triplet_loss_margin=0.1"

--- a/examples/mnist/train.go
+++ b/examples/mnist/train.go
@@ -135,6 +135,7 @@ func TrainModel(ctx *context.Context, dataDir, checkpointPath string, paramsSet 
 
 	fmt.Printf("Training %s model:\n", modelType)
 	backend := backends.New()
+	fmt.Printf("Backend %s: %s\n", backend.Name(), backend.Description())
 
 	lossFn, err := losses.LossFromContext(ctx)
 	if err != nil {

--- a/experimental/README.md
+++ b/experimental/README.md
@@ -1,7 +1,0 @@
-## Experimental Packages
-
-This directory holds experimental extensions that haven't made yet
-to the main API. 
-
-The idea here is that they don't stay long (>1 year?) here: either they are proven useful and moved
-to the main library, or if not proven generically useful, the are moved out to another repository.

--- a/internal/cmd/simplego_dispatcher/main.go
+++ b/internal/cmd/simplego_dispatcher/main.go
@@ -57,6 +57,8 @@ var (
 			{"scatterDTypeMap", "execScatterGeneric", makeDTypes(true, true, true, true, false)},
 			{"dereferenceIntsDTypeMap", "dereferenceIntsGeneric", makeDTypes(true, true, false, false, false)},
 			{"sliceDTypeMap", "execSliceGeneric", makeDTypes(true, true, true, true, true)},
+			{"argMinMaxDTypeMap", "execArgMinMaxGeneric", makeDTypes(true, true, true, false, false)},
+			{"argMinMaxCopyIntsDTypeMap", "buildArgMinMaxCopyIntsFn", makeDTypes(true, true, false, false, false)},
 		},
 		PairMaps: []MapPairInfo{
 			// Various ConvertDType instantiations.

